"It has been 0 days since the last incident" sign

Jacquerel opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

![image](https://user-images.githubusercontent.com/7483112/192075850-19cf6035-9e86-4367-a8b1-961718bf65b7.png)

Adds a fun sign to various places in engineering which tracks how many days (read: rounds) it has been since last time the engine delaminated.

![image](https://user-images.githubusercontent.com/7483112/192075887-4f168d36-10f1-4a8f-979b-145fedd6434e.png)
Example image here shows an unrealistically large number for testing purposes.

![image](https://user-images.githubusercontent.com/7483112/192075902-318bd228-f7bb-40ab-9699-689c8145014d.png)
It ticks up by one at the end of every round and resets to zero on rounds where the engine blows up (or technically to -1, then ticks up to 0 because the round ended).
Not much more to say about it.

## Why It's Good For The Game

I'll be honest I don't have any deep justification for this one except that I think it's cute.
I can make up something about how I think it might encourage people to try not to blow up the engine but I don't believe that nor would I want them to stop.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: Nanotrasen has started tracking exactly how often you're blowing up the supermatter engine. Surely they can't be sending as many new crystals over there as their financial reports imply.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Jacquerel committed:
Adds a sign which tracks how many rounds it has been since the supermatter blew itself up.

Jacquerel committed:
Fixes a race condition.
Actually adds these to maps.
Makes them unwrenchable, because why wouldn't they be.

JohnFulpWillard commented:
I thought Poly's age was the delam incident counter

san7890 commented:
> I thought Poly's age was the delam incident counter

People take Poly with them sometimes

Kapu1178 commented:
make it constructable

LemonInTheDark commented:
This will fail checks, clean up your global list

Mickyan commented:
This is so cool we need one for total station destruction too (nuke/blob/etc.)

LemonInTheDark commented on code/game/objects/structures/signs/signs_interactive.dm:
```diff
@@ -24,3 +24,50 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/sign/calendar, 32)
+	else
+		update_count(SSpersistence.rounds_since_engine_exploded)
+
+/obj/structure/sign/delamination_counter/proc/update_count(new_count)
+	since_last = min(new_count, 99)
+	build_overlays()
+
+/obj/structure/sign/delamination_counter/proc/build_overlays()
+	overlays = list()
+
+	var/ones = since_last % 10
+	var/mutable_appearance/ones_overlay = mutable_appearance('icons/obj/signs.dmi', "days_[ones]")
+	ones_overlay.pixel_x = 4
+	overlays += ones_overlay
+
+	var/tens = (since_last / 10) % 10
+	var/mutable_appearance/tens_overlay = mutable_appearance('icons/obj/signs.dmi', "days_[tens]")
+	tens_overlay.pixel_x = -5
+	overlays += tens_overlay
```
This breaks emissive blocker overlays
Why not just use update_overlays/update_appearance?

msgerbs commented:
> I thought Poly's age was the delam incident counter

Imagine not killing Poly every round

TheBoondock commented:
Now each server can compete for longest days without incidents

LemonInTheDark commented on code/game/objects/structures/signs/signs_interactive.dm:
```diff
@@ -24,3 +24,50 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/sign/calendar, 32)
 		for(var/holidayname in SSevents.holidays)
 			. += span_info("[holidayname]")
+
+GLOBAL_LIST_EMPTY(map_delamination_counters)
+
+/obj/structure/sign/delamination_counter
+	name = "delamination counter"
+	desc = "A pair of flip signs describe how long it's been since the last delamination incident."
+	icon_state = "days_since_explosion"
+	var/since_last = 0
+
+MAPPING_DIRECTIONAL_HELPERS(/obj/structure/sign/delamination_counter, 32)
+
+/obj/structure/sign/delamination_counter/Initialize(mapload)
+	. = ..()
+	if (mapload)
+		GLOB.map_delamination_counters += src
+	else
+		update_count(SSpersistence.rounds_since_engine_exploded)
```
Ah, that's why it passed checks. damn mapload
I know the effect of this is nill, but could you clear out this list on deletion? sooths my neurotic brain

Acer202 commented:
I kinda think it should be in the engine room! Let all the engineers see it. Someone suggests something crazy? Point to the sign. Pfft

zeskorion commented:
does this only go up to 99? have you that little faith?

OrionTheFox commented:
> does this only go up to 99? have you that little faith?

That actually makes me wonder, what happens if the count by SOME miracle goes above 99?

Fikou commented:
add one for chem lab explosions

Jacquerel committed:
Adds delamination counter to constructible signs list.
Overlays are now added in a way which doesn't delete any added from somewhere else.

Jacquerel commented:
> I kinda think it should be in the engine room! Let all the engineers see it. Someone suggests something crazy? Point to the sign. Pfft

There should be one which is reasonably visible to all engineers on every map.

I am not planning on adding other persistent signs in this PR because I am a little worried about oversaturation (and some of the suggestions are harder to detect as events than delamination, which is pretty easy) but it shouldn't be hard to add more, though in that case I would recommend refactoring how the data is saved rather than making one individual file per type of sign.

Elli-Skala commented:
> does this only go up to 99? have you that little faith?

Yes

Wallemations commented:
Are you able to put more than one on each map and have them all sync up or would that make the day counter go up faster

Jacquerel commented:
> Are you able to put more than one on each map and have them all sync up or would that make the day counter go up faster

You can have as many in one map as you want (I put two in most of them), and they're constructible during a round.
The signs don't independently work out the count to display, they ask a subsystem about it, so it won't effect any calculation.

vincentiusvin commented:
You might want to consider adding an is_main_engine check so the sms in syndie(?)/oldstation/wherever else wont trigger it.

JohnFulpWillard commented on code/game/objects/structures/signs/signs_interactive.dm:
```diff
@@ -24,3 +24,56 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/sign/calendar, 32)
+/obj/structure/sign/delamination_counter/update_overlays()
+	. = ..()
+
+	var/ones = since_last % 10
```
Would it not be better to just use maptext here?

Jacquerel replied:
After a bit of messing about, yes, but I think the 5px smallfonts characters are a bit uglier than the sprite overlays as they're a pixel thinner.
![image](https://user-images.githubusercontent.com/7483112/192600479-93b8763c-e2b3-4e2d-b71b-51c4cc38e9f3.png)

If adding 10 overlays to bloat a dmi file is more of a concern than the fact it looks a little worse, I can switch to maptext.

Jacquerel committed:
Only modify counter for main engine.

Jacquerel committed:
Rerun tram test

LemonInTheDark commented:
does it go to a glowing animated rainbow infinite symbol if you go past 99 rounds

Mothblocks commented:
How is this a Mega-Mega-Map-Reworks its one object

Mothblocks commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -460,3 +463,19 @@ SUBSYSTEM_DEF(persistence)
+
+/datum/controller/subsystem/persistence/proc/save_delamination_counter()
+	var/path = "data/rounds_since_delamination.txt"
+	if (fexists(path))
+		fdel(path)
+	text2file("[rounds_since_engine_exploded + 1]", path)
```
```suggestion
	rustg_file_write("[rounds_since_engine_exploded + 1]", path)
```

This will write the file in one pass

Mothblocks commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -460,3 +463,19 @@ SUBSYSTEM_DEF(persistence)
+	if (!fexists(path))
+		return
+	rounds_since_engine_exploded = text2num(file2text(path))
+	for (var/obj/structure/sign/delamination_counter/sign in GLOB.map_delamination_counters)
```
```suggestion
	for (var/obj/structure/sign/delamination_counter/sign as anything in GLOB.map_delamination_counters)
```

Mothblocks commented on code/controllers/subsystem/persistence.dm:
```diff
@@ -460,3 +463,19 @@ SUBSYSTEM_DEF(persistence)
+
+
+/datum/controller/subsystem/persistence/proc/load_delamination_counter()
+	var/path = "data/rounds_since_delamination.txt"
```
Make this a define instead of copying and pasting it in both places.

Mothblocks commented on code/game/objects/structures/signs/signs_interactive.dm:
```diff
@@ -24,3 +24,56 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/sign/calendar, 32)
 		for(var/holidayname in SSevents.holidays)
 			. += span_info("[holidayname]")
+
+GLOBAL_LIST_EMPTY(map_delamination_counters)
```
Doc-comment since two places use it

JohnFulpWillard commented:
It affects all maps, it’s very useful for downstream mappers to keep track of everything they need to add to their own maps.

Mothblocks commented on code/game/objects/structures/signs/signs_interactive.dm:
```diff
@@ -24,3 +24,56 @@ MAPPING_DIRECTIONAL_HELPERS(/obj/structure/sign/calendar, 32)
+
+/obj/structure/sign/delamination_counter/examine(mob/user)
+	. = ..()
+	. += span_info("It has been [since_last] days since the last delamination event at a Nanotrasen facility.")
```
```suggestion
	. += span_info("It has been [since_last] day\s since the last delamination event at a Nanotrasen facility.")
```

Jacquerel committed:
Adds doc comment to global list.
Replaces copied path with define.
Sign formats one day nicely now.
Use alternate file loading method.

Mothblocks commented:
Undraft if ready.

ATH1909 commented:
> and resets to zero on rounds where the engine blows up (or technically to -1, then ticks up to 0 because the round ended).

shouldn't it start at 1 the next round, since it'll have been one day (round) since the engine blew up?

Jacquerel committed:
Signs now live countdown to 0 when the engine goes up, resetting to 1 upon the next round.

Jacquerel commented:
> > and resets to zero on rounds where the engine blows up (or technically to -1, then ticks up to 0 because the round ended).
> 
> shouldn't it start at 1 the next round, since it'll have been one day (round) since the engine blew up?

After thinking it for a while, as much as I like punishing people by having 0 stick around for a whole round I think resetting all signs to 0 when it blows up is probably funnier.
Does require a bit of a change to the global list but as far as I can tell it is cleaning up properly.
